### ISSUE TITLE
Training differences between PyTorch and Tensorflow versions of Ludwig

### ISSUE THREAD

**Issue by dantreiman** [ludwig-ai/ludwig#1613]
Train a ludwig model using concat combiner on Titanic, Forest Cover, or Mushroom Edibility.

The loss curves as well as final results are significantly different.

Desired behavior:
Given the same dataset and config, models trained with TF version of Ludwig vs. Torch may not produce identical results but they should be very similar.

For example, here are the results of running `examples/titanic/multiple_model_training.py` on both Ludwig 0.4 and Ludwig 0.5dev (e065e03c63c6f43a229a2af92ac4cfedaf2b569e)

Ludwig 0.4
![loss_titanic_ludwig0 4](https://user-images.githubusercontent.com/687280/146861212-51c32e4a-78ea-43bc-abc2-8cf37aa97590.png)
![accuracy_titanic_ludwig0 4](https://user-images.githubusercontent.com/687280/146861418-cdef8717-0f27-4070-8ae9-8a08a84a1c90.png)

Ludwig 0.5dev
![loss_titanic_luswig0 5dev](https://user-images.githubusercontent.com/687280/146861222-6df5bc18-5ed5-494d-af6f-4ac689390958.png)
![accuracy_titanic_ludwig0 5dev](https://user-images.githubusercontent.com/687280/146861431-cf940c29-370c-4c47-93b4-dec011cdf97a.png)

**Comment by w4nderlust** [comment]
@dantreiman One thing I would make sure to confirm is first of all the optimizer. Please doublecheck that the same optimizer, with the same parameters is used (also doublecheck the parameters themselves, as sometimes they may be one the flip of the other, like dropout -> dropout_rate). If big differences still persist, I would try using a regular sgd optimizer, in my experience Adam behaves VERY differently in TF and PT (in TF it converges super fast but is really unstable and bouncy, in PT it seems more stable and slower to converge, ut it's just anecdotal). Finally, I would also doublecheck initializations, if they happen correctly for all modules in PT and if the ranges and distributions of the weights after initialization are roughly similar.

**Comment by w4nderlust** [comment]
Anyway, results look different, in particular model1 training, but it's difficult o set a bar really. I'd say the goals here should be to:
1. make sure there are no errors in model building, loss computation (for instance regularization and other special losses like the ones in tablet), optimization and initialization
2. make sure we are not losing performance
3. make sure that in the case where we do, maybe wiggling a bit parameters and config compensates it

Although it should not be the case, little implementation discrepancies between the frameworks may result in different outcomes, and we should be ok with that, but also try to limit issues. Does this make sense to you @dantreiman ?

**Comment by dantreiman** [comment]
Good ideas!  I added a config that trains with SGD, learning rate 0.001, no weight decay or regularization.  Results below.


One thing I notice is that model2 uses zscore normalization in the preprocessing of a few of its features.  model 1 does not have feature normalization.  I wonder if something is different in the working on feature preprocessing or batch norm - since the model with normalized features has the least divergence between torch and tensorflow.



Ludwig 0.4:

![learning_curves_tf_accuracy](https://user-images.githubusercontent.com/687280/146981603-c81543f6-9e9d-42c9-98cb-19d08a5eae2c.png)
![learning_curves_tf_loss](https://user-images.githubusercontent.com/687280/146981606-02373f1c-c2b2-44e4-8f1e-12181a6166f4.png)


Ludwig 0.5dev:

![learning_curves_torch_accuracy](https://user-images.githubusercontent.com/687280/146981621-c466257c-fcd8-4d27-a2de-04eec8dc1220.png)
![learning_curves_torch_loss](https://user-images.githubusercontent.com/687280/146981625-38432866-8caf-446a-b7c1-78de1312aaa5.png)

**Comment by dantreiman** [comment]
Zooming in on model 1 only:

Train Stats:  PyTorch orange, tensorflow red:
<img width="1421" alt="Screen Shot 2022-01-06 at 3 06 02 PM" src="https://user-images.githubusercontent.com/687280/148465005-38b4f7f5-40b2-41ed-82a1-77049d15e631.png">


Test Stats:  PyTorch red, Tensorflow blue.
<img width="1043" alt="Screen Shot 2022-01-06 at 2 43 17 PM" src="https://user-images.githubusercontent.com/687280/148462742-ddd19723-952e-4f4b-bd1d-c20fb6222826.png">

Note:

- Step training loss is a pretty good match, but epoch loss is not.  Did something change in the way we aggregate or display the loss?
- Accuracy is a close match, but AUC is not.  Why?  Perhaps we should plot the final ROC curves.

**Comment by dantreiman** [comment]
#1659 Fixes the chart above.  Turns out epoch loss running average was not being reset between epochs, so the red and orange lines of epoch loss were cumulative averges.
